Add dts profile diff

m-iwanicki opened the pull request:
Related to https://github.com/Dasharo/dts-scripts/pull/97 (needs those changes to work).
UEFI Update profile will need to be generated again because of https://github.com/Dasharo/dts-scripts/commit/171d8decee567e3c55a953c1b28973f0a0a66249

This PR adds basic support for profile checking along with implementing those checks for ODROID H4 workflows.

Comparing profiles is done in TRY block: https://github.com/Dasharo/open-source-firmware-validation/blob/4e0efd9e20e442c64f970906f1bac8ef68d29f05/dts/dts-e2e.robot#L169 in test teardown only if the profile for the test exists.
Maybe adding separate variable in platform config to decide whether profile has to be checked for that workflow would be good idea just in case file or test is renamed. In the current state test would succeed, because it would skip profile checking if it couldn't find reference file.

m-iwanicki commented on dts/profiles/msi-pro-z790-p-ddr5 UEFI->Heads Transition - DPP.profile:
```diff
@@ -0,0 +1,41 @@
+dmidecode -s system-manufacturer 0
```
I'm not sure if this profile is good as after transition I wasn't able to boot into Heads. There was no output on PiKVM and the only output on serial was:

```
Cannot locate primary CBFS
```

DaniilKl replied:
Have you discussed it with Dasharo Team? If the behavior is not expected then the profile should not be trusted.

m-iwanicki replied:
No

DaniilKl replied:
Is the issue still appears? If yes - please add a comment to the profile referencing the issue on `Dasharo/dasharo-issues`. That way the test will fail because of the comment in the profile and we will know what the fail is related to.

m-iwanicki replied:
https://github.com/Dasharo/open-source-firmware-validation/pull/999/commits/4594793ba306db8d20fc9a22e368550c0be735c8

m-iwanicki commented:
I cherry-picked commit from this PR: https://github.com/Dasharo/open-source-firmware-validation/pull/647 to fix https://github.com/Dasharo/dasharo-issues/issues/1526

m-iwanicki commented on dts/profiles/msi-pro-z690-a-ddr5 UEFI->Heads Transition - DPP.profile:
```diff
@@ -0,0 +1,42 @@
+dmidecode -s system-manufacturer 0
```
I couldn't verify if transition worked correctly likely due to https://github.com/Dasharo/dasharo-issues/issues/1533.
I also couldn't get any video output after flashing original fw so maybe there is something wrong with PiKVM

DaniilKl replied:
Is the issue still appears? If yes - please add a comment to the profile referencing the issue on `Dasharo/dasharo-issues`. That way the test will fail because of the comment in the profile and we will know what the fail is related to.

m-iwanicki replied:
https://github.com/Dasharo/open-source-firmware-validation/pull/999/commits/4594793ba306db8d20fc9a22e368550c0be735c8

DaniilKl commented on .gitignore:
```diff
@@ -24,3 +24,5 @@ qemu-data/
 ipxe/
 /logs
+dts/dts-gen-profiles/
+UEFI
```
Why ignore UEFI?

m-iwanicki replied:
Because for some reason it was created when running tests.

DaniilKl reviewed (changes requested):
Additionally add the variable that signals that the platform should have the profile.

DaniilKl commented:
Looks ok. Lets check the tests results.

m-iwanicki commented:
@DaniilKl partial results:

Needs to be fixed:
- odroid - enable profile verification, fixed: https://github.com/Dasharo/open-source-firmware-validation/pull/999/commits/436831c038126fd9435dd27f07eada5d18d99190
- msi and novacustom heads transition - heads transition needs to use https://github.com/Dasharo/open-source-firmware-validation/blob/436831c038126fd9435dd27f07eada5d18d99190/lib/dts-lib.robot#L305 or similar to deal with unflashable ME
- msi uefi update - most likely some platform-config tweaking.

```
==============================================================================
Dts-E2E
==============================================================================
Create tests                                                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E001: novacustom-ns50mu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E002: novacustom-ns50mu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E003: novacustom-ns50pu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E004: novacustom-ns50pu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E005: novacustom-ns70mu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E006: novacustom-ns70mu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E007: novacustom-ns70pu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E008: novacustom-ns70pu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E009: novacustom-nv41mb Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E010: novacustom-nv41mb UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E011: novacustom-nv41mz Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E012: novacustom-nv41mz UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E013: novacustom-v540tnd Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E014: novacustom-v540tnd UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E015: novacustom-v560tnd Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E016: novacustom-v560tnd UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E017: novacustom-v560tne Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E018: novacustom-v560tne UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E019: odroid-h4-plus Initial Deployment - DPP                       | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E020: odroid-h4-plus UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E021: odroid-h4-plus Dasharo (coreboot+UEFI) to Dasharo (Slim Bo... | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E022: odroid-h4-plus Dasharo (Slim Bootloader+UEFI) Initial Depl... | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E023: pcengines-apu2 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E024: pcengines-apu2 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E025: pcengines-apu2 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E026: pcengines-apu3 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E027: pcengines-apu3 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E028: pcengines-apu3 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E029: pcengines-apu4 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E030: pcengines-apu4 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E031: pcengines-apu4 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E032: pcengines-apu6 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E033: pcengines-apu6 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E034: pcengines-apu6 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E035: msi-pro-z690-a-ddr5 Initial Deployment - DCR                  | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E036: msi-pro-z690-a-ddr5 Initial Deployment - DPP                  | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E037: msi-pro-z690-a-ddr5 UEFI Update - DCR                         | PASS |
------------------------------------------------------------------------------
E2E038: msi-pro-z690-a-ddr5 UEFI Update - DPP                         | PASS |
------------------------------------------------------------------------------
E2E039: msi-pro-z690-a-ddr5 UEFI->Heads Transition - DPP              | FAIL |
Teardown failed:
Profiles are not identical!: 1 != 0
------------------------------------------------------------------------------
E2E040: msi-pro-z690-a-wifi-ddr4 Initial Deployment - DCR             | PASS |
------------------------------------------------------------------------------
E2E041: msi-pro-z690-a-wifi-ddr4 Initial Deployment - DPP             | PASS |
------------------------------------------------------------------------------
E2E042: msi-pro-z690-a-wifi-ddr4 UEFI Update - DCR                    | PASS |
------------------------------------------------------------------------------
E2E043: msi-pro-z690-a-wifi-ddr4 UEFI Update - DPP                    | PASS |
------------------------------------------------------------------------------
E2E044: msi-pro-z690-a-wifi-ddr4 UEFI->Heads Transition - DPP         | FAIL |
No match found for 'Successfully switched to Dasharo Heads firmware' in 5 minutes
Output:

Updating Dasharo firmware...
This may take several minutes. Please be patient and do not
power off your computer or touch the keyboard!
Backing up firmware configuration... Failed! Default settings will be used.
Done.
Can not determine if ME is disabled, assuming enabled.
Checking flash layout.
The firmware binary contains Management Engine (ME), but ME is not disabled!

Flashing ME when not in disabled state may cause unexpected power management issues.
Recovering from such state may require removal of AC power supply and resetting CMOS battery.
Keeping an older version of ME may cause a CPU to perform less efficient, e.g. if upgraded the CPU to a newer generation.
You have been warned.

Skip ME flashing and proceed with BIOS/firmware flashing/updating? (Y|n) .
------------------------------------------------------------------------------
E2E045: msi-pro-z790-p-ddr5 Initial Deployment - DPP                  | PASS |
------------------------------------------------------------------------------
E2E046: msi-pro-z790-p-ddr5 UEFI Update - DPP                         | FAIL |
Teardown failed:
Profiles are not identical!: 1 != 0
------------------------------------------------------------------------------
E2E047: msi-pro-z790-p-ddr5 UEFI->Heads Transition - DPP              | FAIL |
Teardown failed:
Profiles are not identical!: 1 != 0
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E048: novacustom-nv41pz Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E049: novacustom-nv41pz UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
```

m-iwanicki commented:
@DaniilKl results:

<details><summary>61 tests, 57 passed, 4 failed</summary>
<p>

```text
==============================================================================
Dts-E2E
==============================================================================
Create tests                                                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E001: novacustom-ns50mu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E002: novacustom-ns50mu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E003: novacustom-ns50pu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E004: novacustom-ns50pu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E005: novacustom-ns70mu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E006: novacustom-ns70mu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E007: novacustom-ns70pu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E008: novacustom-ns70pu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E009: novacustom-nv41mb Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E010: novacustom-nv41mb UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E011: novacustom-nv41mz Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E012: novacustom-nv41mz UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E013: novacustom-v540tnd Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E014: novacustom-v540tnd UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E015: novacustom-v560tnd Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E016: novacustom-v560tnd UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E017: novacustom-v560tne Initial Deployment - DCR                   | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E018: novacustom-v560tne UEFI Update - DCR                          | PASS |
------------------------------------------------------------------------------
E2E019: odroid-h4-plus Initial Deployment - DPP                       | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E020: odroid-h4-plus UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E021: odroid-h4-plus Dasharo (coreboot+UEFI) to Dasharo (Slim Bo... | PASS |
------------------------------------------------------------------------------
E2E022: odroid-h4-plus Dasharo (Slim Bootloader+UEFI) Initial Depl... | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E023: pcengines-apu2 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E024: pcengines-apu2 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E025: pcengines-apu2 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E026: pcengines-apu3 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E027: pcengines-apu3 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E028: pcengines-apu3 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E029: pcengines-apu4 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E030: pcengines-apu4 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E031: pcengines-apu4 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E032: pcengines-apu6 UEFI Update - DPP                              | PASS |
------------------------------------------------------------------------------
E2E033: pcengines-apu6 SeaBIOS Update - DPP                           | PASS |
------------------------------------------------------------------------------
E2E034: pcengines-apu6 SeaBIOS->UEFI Transition - DPP                 | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E035: msi-pro-z690-a-ddr5 Initial Deployment - DCR                  | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E036: msi-pro-z690-a-ddr5 Initial Deployment - DPP                  | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E037: msi-pro-z690-a-ddr5 UEFI Update - DCR                         | PASS |
------------------------------------------------------------------------------
E2E038: msi-pro-z690-a-ddr5 UEFI Update - DPP                         | PASS |
------------------------------------------------------------------------------
E2E039: msi-pro-z690-a-ddr5 UEFI->Heads Transition - DPP              | PASS |
------------------------------------------------------------------------------
E2E040: msi-pro-z690-a-wifi-ddr4 Initial Deployment - DCR             | PASS |
------------------------------------------------------------------------------
E2E041: msi-pro-z690-a-wifi-ddr4 Initial Deployment - DPP             | PASS |
------------------------------------------------------------------------------
E2E042: msi-pro-z690-a-wifi-ddr4 UEFI Update - DCR                    | PASS |
------------------------------------------------------------------------------
E2E043: msi-pro-z690-a-wifi-ddr4 UEFI Update - DPP                    | PASS |
------------------------------------------------------------------------------
E2E044: msi-pro-z690-a-wifi-ddr4 UEFI->Heads Transition - DPP         | FAIL |
Teardown failed:
Profiles are not identical!: 1 != 0
------------------------------------------------------------------------------
E2E045: msi-pro-z790-p-ddr5 Initial Deployment - DPP                  | PASS |
------------------------------------------------------------------------------
E2E046: msi-pro-z790-p-ddr5 UEFI Update - DPP                         | PASS |
------------------------------------------------------------------------------
E2E047: msi-pro-z790-p-ddr5 UEFI->Heads Transition - DPP              | FAIL |
Teardown failed:
Profiles are not identical!: 1 != 0
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E048: novacustom-nv41pz Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
E2E049: novacustom-nv41pz UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
E2E050: novacustom-nv41pz UEFI->Heads Transition - DPP                | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E051: novacustom-v540tu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E052: novacustom-v540tu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
E2E053: novacustom-v540tu UEFI->Heads Transition - DPP                | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E054: novacustom-v560tu Initial Deployment - DCR                    | PASS |
------------------------------------------------------------------------------
[ WARN ] Workflow isn't configured for profile verification.
E2E055: novacustom-v560tu UEFI Update - DCR                           | PASS |
------------------------------------------------------------------------------
E2E056: novacustom-v560tu UEFI->Heads Transition - DPP                | PASS |
------------------------------------------------------------------------------
E2E057: optiplex-7010 Initial Deployment - DPP                        | FAIL |
Teardown failed:
Several failures occurred:

1) File '/home/miwanicki/projects/open-source-firmware-validation/dts/profiles/optiplex-7010 Initial Deployment - DPP.profile' does not exist.

2) Profiles are not identical!: 2 != 0
------------------------------------------------------------------------------
E2E058: optiplex-7010 UEFI Update - DPP                               | PASS |
------------------------------------------------------------------------------
E2E059: optiplex-9010 Initial Deployment - DPP                        | FAIL |
Teardown failed:
Several failures occurred:

1) File '/home/miwanicki/projects/open-source-firmware-validation/dts/profiles/optiplex-9010 Initial Deployment - DPP.profile' does not exist.

2) Profiles are not identical!: 2 != 0
------------------------------------------------------------------------------
E2E060: optiplex-9010 UEFI Update - DPP                               | PASS |
------------------------------------------------------------------------------
Dts-E2E                                                               | FAIL |
61 tests, 57 passed, 4 failed
==============================================================================
```

</p>
</details>

2 failures are expected (1 non-working heads transition on Z790 and 1 unverified heads transition).
2 other failures are due to missing profiles, or rather, wrong profile filename (it's `Initial Deployment` not `Initial deployment`), fixed here: https://github.com/Dasharo/open-source-firmware-validation/pull/999/commits/9a3edd831c2eefbd1e34485d39242f1803695e0e
Even after that profile for optiplex-7010 is different, fixed in 9c51db6257f1124d363c2e7b4796e106ae74858f.
I'll rerun tests one more time to make sure everything works and then squash most of commits in this PR:

```
------------------------------------------------------------------------------
E2E001: optiplex-7010 Initial Deployment - DPP                        | PASS |
------------------------------------------------------------------------------
E2E002: optiplex-9010 Initial Deployment - DPP                        | PASS |
------------------------------------------------------------------------------
```

m-iwanicki commented:
Every test passed except those 2 expected

DaniilKl commented:
@m-iwanicki, please, clean up the commits a bit.

m-iwanicki committed:
dts-e2e: add workflow profiles

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-e2e: modify platform configs so they reflect real workflows

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-e2e: modify e2e tests to verify profile at the end

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts: add tests to generate profiles on real platforms

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
flash.robot: use INTERNAL_PROGRAMMER_CHIPNAME with quotes

Otherwise it will fail if chipname contains space

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
network.robot: Allow optional hash verification in 'Get File From DUT'

This allows us to use this keyword with globs to get multiple files from DUT

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
network.robot: Try to use 'ip a' to get IP if 'hostname -I' fails

On DTS 'hostname' command doesn't have 'I' argument

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
ipxe-run.sh: remove root=/dev/nfs to fix dnf install

With this argument installing DTS extensions fails as dnf thinks there
is not enough free space on rootfs

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
docs: dts-tests.md: split long robot command into multiple lines

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>
